File: 4_kyu/Human_readable_duration_format.kt
object TimeFormatter {

    fun formatDuration(seconds: Int): String {
        val nameAndCount = arrayOf( Pair("second",60), Pair("minute",60), Pair("hour",24), Pair("day",365), Pair("year",Int.MAX_VALUE) )
        val acc = ArrayList<String>( nameAndCount.size )
        var n = seconds
        for( nc in nameAndCount ) {
            val i = n % nc.second
            when {
                i == 1 -> acc.add( 0, "1 ${nc.first}" )
                i  > 1 -> acc.add( 0, "${i} ${nc.first}s" )
            }
            n /= nc.second
        }
        return when(acc.size) {
            0 -> "now"
            1 -> acc[0]
            2 -> "${acc[0]} and ${acc[1]}"
            else -> acc.dropLast(2).joinToString() + ", ${acc[acc.lastIndex-1]} and ${acc[acc.lastIndex]}"
        }
    }
}

File: 4_kyu/Human_readable_duration_format.rs
// Used units in the output string.
// Each item: (count up to next unit, name of the unit without plural "s").
static UNITS: [(u64, &str); 5] = [
    (60, "second"),
    (60, "minute"),
    (24, "hour"),
    (365, "day"),
    (u64::MAX, "year"),
];

fn format_duration(seconds: u64) -> String {
    // 1) Split the seconds into parts
    let mut parts = Vec::<String>::with_capacity(UNITS.len());
    let mut rest = seconds;
    for (count, name) in UNITS {
        let n = rest % count;
        rest /= count;
        if n > 0 {
            parts.push(format!("{} {}{}", n, name, if n > 1 { "s" } else { "" }));
        }
    }
    // 2) Combine the parts wtih biggest unit first
    parts.reverse();
    match parts.len() {
        0 => String::from("now"),
        1 => parts.pop().unwrap(),
        n => format!("{} and {}", parts[0..n - 1].join(", "), parts[n - 1]),
    }
}
